### PR TITLE
Add -fvisibility=hidden to macOS dynamic framework and export required symbols

### LINKED ISSUE #7281: Many visibility warnings building macOS SDK
As of 84038832e45c2445b43b39ccb2f9e29c8646dac0 on master, a clean build of the macOS SDK in Xcode 8.2 results in 136 warnings of the following form:

```
ld: warning: direct access in function 'mbgl::Tileset const& mapbox::util::variant<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, mbgl::Tileset>::get<mbgl::Tileset, (void*)0>() const' from file '/Users/mxn/hub/mapbox-gl-native/build/macos/Debug/libmbgl-core.a(offline_download.o)' to global weak symbol 'typeinfo for mapbox::util::bad_variant_access' from file '/Users/mxn/hub/mapbox-gl-native/build/macos/macos.build/Debug/dynamic.build/Objects-normal/x86_64/NSArray+MGLAdditions.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
```

`GCC_SYMBOLS_PRIVATE_EXTERN` is `YES` for the dynamic target in macos.xcodeproj (whereas it’s `NO` for the dynamic and static targets in ios.xcodeproj), but setting it to `NO` doesn’t address the issue.

This issue doesn’t reproduce on the release-ios-v3.4.0 branch.

/cc @kkaefer

### PR DESCRIPTION
**Pull request by kkaefer**
Reroll of #7326, which apparently has broken CI statuses.

Fixes #7326
Fixes #7281

### CODE REVIEW

**Comment by kkaefer** [comment]
Grr, this seems to be broken as well...

**Comment by kkaefer** [comment]
Let's try approving this to see if we can merge it.

File: platform/darwin/src/MGLAccountManager.h [hunk 1]
```diff
@@ -1,11 +1,12 @@
-#import <Foundation/Foundation.h>
```
Comment (1ec5) [inline platform/darwin/src/MGLAccountManager.h]: Apparently we need these import statements or else Swift declarations in the generated documentation lack specific types: realm/jazzy#609. Importing required system frameworks is good practice anyways, unless you use a precompiled header that imports them.

(Sorry, I started this review a couple weeks ago but forgot to publish it.)

Comment (kkaefer) [inline platform/darwin/src/MGLAccountManager.h]: `MGLTypes.h` imports `Foundation.h`

**Comment by kkaefer** [comment]
> What about protocols and categories? Do they need to be exported as well?

No, we only need to export compiled code, but protocols/categories are a header-only feature.

File: platform/darwin/src/MGLAccountManager.h (hunk 1 above)
Comment (1ec5) [inline platform/darwin/src/MGLAccountManager.h]: As mentioned in that jazzy ticket, this header itself must import Foundation.h. It's unfortunate, but that shouldn't pose a problem for the visibility changes you're making, should it?

File: platform/darwin/src/MGLBackgroundStyleLayer.h [hunk 2]
```diff
@@ -12,6 +12,7 @@ NS_ASSUME_NONNULL_BEGIN
  `style` and obtain the background layer using the `-[MGLStyle layerWithIdentifier:]` 
  method and passing `background` for the identifier. 
  */
+MGL_EXPORT
```
Comment (1ec5) [inline platform/darwin/src/MGLBackgroundStyleLayer.h]: This change will get blown away as soon as anyone invokes `make style-code`. Make this change in platform/darwin/src/MGLStyleLayer.h.ejs and run `make style-code` to propagate the change to all the affected headers.

File: platform/darwin/src/MGLAccountManager.h (hunk 1 above)
Comment (1ec5) [inline platform/darwin/src/MGLAccountManager.h]: I’m not sure why, but we don’t seem to be affected by realm/jazzy#609 even with these changes. Running `make xdocument` turns up some broken Swift declarations, but that happens on master too. (@incanus may have a stronger opinion than I about explicitly importing frameworks in headers. My concern was limited to jazzy output.)

**Comment by 1ec5** [comment]
>> What about protocols and categories? Do they need to be exported as well?
> 
> No, we only need to export compiled code, but protocols/categories are a header-only feature.

Protocols are header-only in Objective-C, but categories may or may not have an implementation in a translation unit separate from the categorized class. (Otherwise, it’s a class extension.) See platform/darwin/src/NSValue+MGLAdditions.m for example.

**Comment by kkaefer** [comment]
I've rebuilt this PR based on SourceKitten's output.

Here are the changes:
- Moved the `MGL_EXPORT` macro to a `MGLFoundation.h` to mirror other Cocoa frameworks usage of the `*_EXPORT` macro.
- Left includes for `Foundation.h` in place
- Added a script that checks all exported symbols for presence of a `MGL_EXPORT` macro
- Added the automated script to `make xpackage`

These symbols are not exposed in the public headers, but the Xcode CI unit tests include the private headers and need them in the library. These types of symbols are not checked by the above script, but we'll catch those easily because the test will fail to run on CI.

- `MGLAttributionButton`
- `MGLAltitudeForZoomLevel`
- `MGLZoomLevelForAltitude`
- `MGLFeaturesFromMBGLFeatures`
- `MGLTileSetFromTileURLTemplates`
- `MGLGeoJSONOptionsFromDictionary`

File: platform/darwin/src/MGLBackgroundStyleLayer.h (hunk 2 above)
Comment (kkaefer) [inline platform/darwin/src/MGLBackgroundStyleLayer.h]: This is fixed

**Comment by kkaefer** [comment]
> Protocols are header-only in Objective-C, but a category may or may not have an implementation in a translation unit separate from the categorized class.

I checked this and it seems they don't need the `MGL_EXPORT` statement and are automatically exported. The CI tests + macOS app uses some of those symbols and can be successfully linked.

File: platform/macos/src/Mapbox.h [hunk 3]
```diff
@@ -1,10 +1,12 @@
 #import <Cocoa/Cocoa.h>
 
+#import "MGLFoundation.h"
+
 /// Project version number for Mapbox.
-FOUNDATION_EXPORT double MapboxVersionNumber;
+FOUNDATION_EXPORT MGL_EXPORT double MapboxVersionNumber;
```
Comment (1ec5) [inline platform/macos/src/Mapbox.h]: Make sure `MapboxVersionNumber` and `MapboxVersionString` are correctly set in the output build. I would hope that Xcode isn’t dependent on the particular syntax used to declare these constants, but you never know.

Comment (1ec5) [inline platform/macos/src/Mapbox.h]: I get this linker error referencing `MapboxVersonNumber` within Swift code:

```
Undefined symbols for architecture x86_64:
  "_MapboxVersionNumber", referenced from:
      Invincibility.ViewController.viewDidLoad () -> () in ViewController.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

File: platform/darwin/src/MGLFoundation.h [hunk 4]
```diff
@@ -0,0 +1,5 @@
+#pragma once
+
+#import <Foundation/Foundation.h>
+
+#define MGL_EXPORT __attribute__((visibility ("default")))
```
Comment (kkaefer) [inline platform/darwin/src/MGLFoundation.h]: As per convention in other system frameworks, it has things that are required in virtually every header file. Most other *Foundation.h files I checked only have the export (or extern) macro.

File: platform/macos/DEVELOPING.md [hunk 5]
```diff
@@ -54,6 +54,7 @@ To add any Objective-C type, constant, or member to the iOS SDK’s public inter
 
 To add an Objective-C class, protocol, category, typedef, enumeration, or global constant to the macOS SDK’s public interface:
 
+1. _(Optional.)_ Add the macro `MGL_EXPORT` prior to the declaration for classes and global constants. To use this macro, include `MGLFoundation.h`. You can check whether all public symbols are exported correctly by running `platform/darwin/scripts/check-public-symbols.js macOS`.
```
Comment (kkaefer) [inline platform/macos/DEVELOPING.md]: Added `make check-public-symbols`

File: platform/macos/src/Mapbox.h (hunk 3 above)
Comment (kkaefer) [inline platform/macos/src/Mapbox.h]: From a C build perspective, this is very weird. `FOUNDATION_EXPORT` is only macro for `extern`, meaning it should be defined in some translation unit (but it's not). Should we manually define this version number?

Comment (kkaefer) [inline platform/macos/src/Mapbox.h]: Looks like Xcode automatically generates a file called `Mapbox_vers.c`

**Comment by kkaefer** [comment]
> can you add a blurb to the macOS changelog if a significant size reduction results from these changes?

There is no size reduction with this PR. This already happened in https://github.com/mapbox/mapbox-gl-native/pull/7130. This one is just a followup that also compiles the Objective-C code with `-fvisibility=hidden` to fix https://github.com/mapbox/mapbox-gl-native/issues/7281.

File: platform/macos/macos.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -1558,7 +1568,7 @@
 				PRODUCT_NAME = Mapbox;
 				SKIP_INSTALL = YES;
 				VERSIONING_SYSTEM = "apple-generic";
-				VERSION_INFO_PREFIX = "";
+				VERSION_INFO_PREFIX = "__attribute__((visibility (\"default\"))) ";
```
Comment (kkaefer) [inline platform/macos/macos.xcodeproj/project.pbxproj]: We're now also adding the visibility change to the auto-generated file.

File: platform/darwin/test/MGLVersionNumber.m [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+@interface MGLVersionTests : XCTestCase
+
+@end
+
+@implementation MGLVersionTests
+
+- (void)testVersionNumber {
+    XCTAssertEqual(1, MapboxVersionNumber);
+}
```
Comment (kkaefer) [inline platform/darwin/test/MGLVersionNumber.m]: Added a test here to check linkage. This is not a Swift-specific issue; it also happened with Objective-C.

File: platform/macos/macos.xcodeproj/project.pbxproj (hunk 6 above)
Comment (1ec5) [inline platform/macos/macos.xcodeproj/project.pbxproj]: Clever hack. (I think this build setting is usually used for Apple to prepend a class prefix like “NS” or “UI” to the variable name.)

File: platform/darwin/test/MGLVersionNumber.m (hunk 7 above)
Comment (1ec5) [inline platform/darwin/test/MGLVersionNumber.m]: Note that `MapboxVersionString` is also available to Objective-C (but not to Swift, owing to its type). I assume it’s also covered by `VERSION_INFO_PREFIX`.

**Comment by 1ec5** [comment]
iOS builds are failing because the iOS version of Mapbox.h doesn’t import MGLFoundation.h:

```
❌  /Users/vagrant/git/platform/darwin/src/MGLAccountManager.h:3:9: include of non-modular header inside framework module 'Mapbox.MGLAccountManager'

#import "MGLFoundation.h"
        ^
```

File: platform/macos/DEVELOPING.md (hunk 5 above)
Comment (1ec5) [inline platform/macos/DEVELOPING.md]: Great. Can you refer to `make check-public-symbols` in both DEVELOPING.md documents? Thanks!

File: platform/darwin/test/MGLVersionNumber.m (hunk 7 above)
Comment (kkaefer) [inline platform/darwin/test/MGLVersionNumber.m]: Yeah, it is.

**Pull request merged**